sdk/src/main/java/com/mapbox/maps/renderer/widget/BitmapWidgetRenderer.kt

```diff
@@ -0,0 +1,326 @@
+      this.put(value)
+    }
+    rewind()
+  }
+
+  override fun onSurfaceChanged(width: Int, height: Int) {
+    surfaceWidth = width
+    surfaceHeight = height
+
+    // transforms from (0,0) - (width, height) in screen pixels
+    // to (-1, -1) - (1, 1) for GL
+    screenMatrix.put(
+      2f / width, 0f, 0f, 0f,
+      0f, -2f / height, 0f, 0f,
+      0f, 0f, 0f, 0f,
+      -1f, 1f, 0f, 1f
+    )
+
+    Matrix.translateM(
+      translateMatrix,
+      0,
+      topX(),
+      leftY(),
+      0f
+    )
+
+    updateVertexBuffer()
+
+    updateMatrix = true
+    needRender = true
+  }
+
+  private fun updateVertexBuffer() {
+    // in pixels, (-bitmapWidth / 2, -bitmapHeight/2) - (bitmapWidth / 2, bitmapHeight/2)
+    vertexPositionBuffer.put(
+      -bitmapWidth / 2f, -bitmapHeight / 2f,
+      -bitmapWidth / 2f, bitmapHeight / 2f,
+      bitmapWidth / 2f, -bitmapHeight / 2f,
+      bitmapWidth / 2f, bitmapHeight / 2f,
+    )
+  }
+
+  private fun leftY() = when (position) {
+    WidgetPosition.BOTTOM_LEFT,
+    WidgetPosition.BOTTOM_RIGHT -> surfaceHeight.toFloat() - bitmapHeight.toFloat() / 2f - marginY
+    WidgetPosition.TOP_LEFT,
+    WidgetPosition.TOP_RIGHT -> marginY + bitmapHeight.toFloat() / 2f
+  }
+
+  private fun topX() = when (position) {
+    WidgetPosition.TOP_RIGHT,
+    WidgetPosition.BOTTOM_RIGHT -> surfaceWidth.toFloat() - bitmapWidth.toFloat() / 2f - marginX
+    WidgetPosition.TOP_LEFT,
+    WidgetPosition.BOTTOM_LEFT -> marginX + bitmapWidth.toFloat() / 2f
+  }
+
+  override fun initialize() {
+    val maxAttrib = IntArray(1)
+    GLES20.glGetIntegerv(GLES20.GL_MAX_VERTEX_ATTRIBS, maxAttrib, 0)
+
+    vertexShader = GlUtils.loadShader(
+      GLES20.GL_VERTEX_SHADER,
+      VERTEX_SHADER_CODE
+    ).also(GlUtils::checkCompileStatus)
+
+    fragmentShader = GlUtils.loadShader(
+      GLES20.GL_FRAGMENT_SHADER,
+      FRAGMENT_SHADER_CODE
+    ).also(GlUtils::checkCompileStatus)
+
+    program = GLES20.glCreateProgram().also { program ->
+      GlUtils.checkError("glCreateProgram")
+
+      GLES20.glAttachShader(program, vertexShader)
+      GlUtils.checkError("glAttachShader")
+
+      GLES20.glAttachShader(program, fragmentShader)
+      GlUtils.checkError("glAttachShader")
+
+      GLES20.glLinkProgram(program)
+      GlUtils.checkError("glLinkProgram")
+    }
+
+    uniformMvpMatrix =
+      GLES20.glGetUniformLocation(program, "uMvpMatrix")
+    GlUtils.checkError("glGetUniformLocation")
+
+    attributeVertexPosition =
+      GLES20.glGetAttribLocation(program, "aPosition")
+    GlUtils.checkError("glGetAttribLocation")
+
+    attributeTexturePosition =
+      GLES20.glGetAttribLocation(program, "aCoordinate")
+    GlUtils.checkError("glGetAttribLocation")
+
+    uniformTexture =
+      GLES20.glGetUniformLocation(program, "uTexture")
+    GlUtils.checkError("glGetUniformLocation")
+
+    needRender = true
+  }
+
+  override fun render() {
+    if (program == 0) {
+      initialize()
```

Comment:
same as comment above - I'd organize code in a way the we def init before rendering